[export] support non-persistent buffers

suo opened the pull request:
Summary:
Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

One weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

As a side effect, this diff tightened up quite a few sloppy  behaviors around state dict handling:
- Tensor attributes were getting promoted to be buffers—bad!
- Tracing through a module not in the children of the root module would add its parameters/buffers to the state dict—bad!

This behavior is unlikely to show up in user code since the model would be totally broken, but did show up in a bunch of tests.

#buildmore

Test Plan: unit tests

Differential Revision: D53314477




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

guangy10 commented:
@suo To fix thoe job [*pull / linux-jammy-py3-clang12-executorch / test (executorch, 1, 1, linux.2xlarge)*](https://github.com/pytorch/pytorch/actions/runs/7744402825/job/21118732168?pr=118876), you can edit the pinned commit [pytorch/pytorch/.ci/docker/ci_commit_pins/executorch.txt](https://github.com/pytorch/pytorch/blob/main/.ci/docker/ci_commit_pins/executorch.txt) to point to the WIP PR #[1802](https://github.com/pytorch/executorch/pull/1802). You can either put the commit hash or the branch name "suo:export-D53314477". CC: @huydhn

guangy10 reviewed (approved):
~~Cool! I see `[linux-jammy-py3-clang12-executorch](https://github.com/pytorch/pytorch/actions/runs/7744402825/job/21118066949?pr=118876#logs)` is green now.~~ No, the test still fails. Let me dig into it a bit.

titaiwangms commented on torch/onnx/_internal/io_adapter.py:
```diff
@@ -611,9 +611,13 @@ def apply(
+            if name in non_persistent_buffers:
+                ordered_buffers.append(model.constants[name])  # type: ignore[union-attr]
+            else:
+                ordered_buffers.append(model.state_dict[name])  # type: ignore[union-attr,index]
```
Hi @suo I noticed in ExportedProgram that although we are lifting `constant_tensors`, and `constants`, they are saved individually, not in state_dict. Are they going to be merged in one dict? As from graph execution perspective, backends would need to get these inputs from different places (state_dict, constant_tensors, constants). Thanks!

suo replied:
No, they will not be merged. The state_dict contains parameters/buffers, non-persistent buffers are not saved at all in eager pytorch and are reinitialized each time. For export we can't do that, so we save non-persistent buffers to the constants dict.

titaiwangms reviewed (approved):
All good in ONNX

suo committed:
[export] support non-persistent buffers (#118876)

Summary:

X-link: https://github.com/pytorch/executorch/pull/1802

Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

One weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

As a side effect, this diff tightened up quite a few sloppy  behaviors around state dict handling:
- Tensor attributes were getting promoted to be buffers—bad!
- Tracing through a module not in the children of the root module would add its parameters/buffers to the state dict—bad!

This behavior is unlikely to show up in user code since the model would be totally broken, but did show up in a bunch of tests.

#buildmore

Test Plan: unit tests

Reviewed By: mergennachin

Differential Revision: D53314477
